Build_Process/GNUMakefiles.md

```diff
@@ -0,0 +1,169 @@
+
+Next up is an important line: `.PHONY: `. Make targets are presumed to output a file of the same name by default. By adding a target as a dependency of the .PHONY target, we tell make that this target is more of a command, and not a real file it should look for. In simple scenarios it serves little purpose, but it's an issue that can catch you by surprise later on. Since phony targets don't have a time associated with them, they are assumed to be always out of date and thus are run everytime they are called.
+
+The `all` and `clean` targets work as you'd expect, building the final output or cleaning the build files. It's worth noting the '@' symbol in front of echo. When at the start of a line, it tells make not to echo the rest of the line to the shell. In this case we're using echo because we want to output text without the shell trying to run it. Therefore we tell make not to output the echo line itself, since echo will already write the following text to the shell.
+
+The line `-rm -r build/` begins with a minus/hyphen. Normally if a command fails (returns a non-zero exit code), make will abort the sequence of commands and display an error. Beginning a line a hyphen tells make to ignore the error code. Make will still tell you an error occured, but it won't stop the executing the make file. In this case this is what we want.
+
+The last two rules tell make how it should create a ``*.c.o` or `*.S.o` file if it needs them. They have a dependency on a file of the same name, but with a different extension (`*.c` or `*.S`). This means make will fail with an error if the source file does not exist, or if we forget to add it to the `SRCS` variables above. We do a protective mkdir, to ensure that the filepath used for output actually exists.
+
+Make has a few built-in symbols that are used through the above example makefile. These are called automatic variables, and are described in the makefile manual.
+
+That's a lot of text! But here we can see an example of a number of make functions being used. This provides a simple, but very flexible build system for a project, even allowing the tools to be swapped out by editing a few lines.
+
+There are other built in functions and symbols that have useful meanings, however discovering them is left as an exercise to the reader.
+
+## Complex Makefile Example (with recursion!)
+What about bigger projects? Well you aren't limited to a single makefile, one makefile can include another one (essentially copy-pasting it into the current file) using the `include` keyword. 
+For example, to include `extras.mk` (.mk is a common extension for non-primary makefiles) into `Makefile` you would add the line somewhere:
+```makefile
```

Comment:
remember the newline when writing codes (is needed by pandoc to render correctly code blocks)